### PR TITLE
Dan cta bus

### PR DESCRIPTION
**Pull request by dan8239**
Hey guys, tested the CTA Bus near me user story via dummy JSON file inputting route 20 and Eastbound. Note that it doesn't work with the 135, but that's due to the CTA app not having any stop predictions for that route. It throws a "none scheduled" error. We'll have to error protect that.

Also note that I added slots to the intent, to account for us needing to specify a direction when calling the bus API.

I have a funeral to attend tomorrow so I cannot meet. Please pull this in without me if at all possible.

### CODE REVIEW

File: .ask/config [hunk 1]
```diff
@@ -1,14 +1,14 @@
 {
   "deploy_settings": {
     "default": {
-      "skill_id": "amzn1.ask.skill.e0929fb0-ad82-43f5-b785-95eee4ddef38",
+      "skill_id": "amzn1.ask.skill.68bc5ea2-f393-4596-a0d8-64f9e1790a09",
```
Comment (baumannalexj) [inline .ask/config]: reminder to revert

File: .ask/config [hunk 2]
```diff
@@ -1,14 +1,14 @@
...
       "was_cloned": false,
       "merge": {
         "manifest": {
           "apis": {
             "custom": {
               "endpoint": {
-                "uri": "chicago-now-dev-main"
+                "uri": "chicago-now-dan"
```
Comment (baumannalexj) [inline .ask/config]: reminder to revert

File: lambda/custom/helpers/DistanceCalculator.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+    var radlat2 = Math.PI * lat2/180;
+    var radlon1 = Math.PI * lon1/180;
+    var radlon2 = Math.PI * lon2/180;
+    var theta = lon1-lon2;
+    var radtheta = Math.PI * theta/180;
+    var dist = Math.sin(radlat1) * Math.sin(radlat2) + Math.cos(radlat1) * Math.cos(radlat2) * Math.cos(radtheta);
+    dist = Math.acos(dist);
+    dist = dist * 180/Math.PI;
+    dist = dist * 60 * 1.1515;
+    if (unit=="K") { dist = dist * 1.609344; }
```
Comment (baumannalexj) [inline lambda/custom/helpers/DistanceCalculator.js]: we can use "switch"  and make the units a const/enum in DistanceCalculator

File: skill.json [hunk 4]
```diff
@@ -3,7 +3,7 @@
     "publishingInformation": {
       "locales": {
         "en-US": {
-          "name": "ChicagoNow"
+          "name": "ChicagoNow-dan"
```
Comment (baumannalexj) [inline skill.json]: reminder to revert

File: lambda/custom/helpers/UTCTimeToEnglish.js [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+    var hour = d.hour == 0 ? 12 : (d.hour > 12 ? d.hour - 12 : d.hour);
+    var min = d.minute;
+    var ampm = d.hour < 12 ? 'AM' : 'PM';
+    var time = hour + ':' + min + ' ' + ampm;
+    return time;
+}
+
+
+var CtaDate = (ctaDate) => {
+    let  ctaObject = {
```
Comment (baumannalexj) [inline lambda/custom/helpers/UTCTimeToEnglish.js]: I'll rename these

File: lambda/custom/handlers/transit/bus/BusHandler.js [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+const BusRepository = require('../../../repositories/transit/CtaBusRepository');
+const util = require('util');
+
+
+const CTABUS_API_KEY = 'mY73pz65XVB4Yc7GYAgqFrHQY';
+const CTABUS_API_DOMAIN = 'http://ctabustracker.com';
+const CTABUS_API_PATH = '/bustime/api/v2/getpredictions';
+
+exports.getBusesForRouteAndStop = (parameters, callback) => {
+    console.log("Entered BusHandler.getBusesForRouteAndStop");
```
Comment (baumannalexj) [inline lambda/custom/handlers/transit/bus/BusHandler.js]: reminder

File: lambda/custom/handlers/transit/bus/BusHandler.js [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+        path: CTABUS_API_PATH,
+        queryParams: {
+            key: CTABUS_API_KEY,
+            rt: parameters.rt,
+            stpid: parameters.stpid,
+            format: "json"
+        }
+        
+    });
+    console.log(url);
```
Comment (baumannalexj) [inline lambda/custom/handlers/transit/bus/BusHandler.js]: reminder

File: lambda/custom/handlers/transit/bus/BusHandler.js [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    request(url,  (error, response, body) => {      
+        console.log("entered request method");
```
Comment (baumannalexj) [inline lambda/custom/handlers/transit/bus/BusHandler.js]: reminder

File: lambda/custom/handlers/transit/bus/BusHandler.js [hunk 9]
```diff
@@ -0,0 +1,49 @@
...
+    });
+};
+
+/*
+Call getLocation, forward longitude and parameters entered to Bus Stop Repository
+*/
+exports.searchBusNearMe = (parameters, callback) => {
+    console.log("BusHandler.searchBusNearMe() routing to LocationHandler");
+    LocationHandler.getLocation(parameters.apiEndpoint, parameters.token, parameters.deviceID, (location) => {
+        console.log("latitude: " + location.latitude);
```
Comment (baumannalexj) [inline lambda/custom/handlers/transit/bus/BusHandler.js]: reminder

File: lambda/custom/repositories/transit/CtaBusRepository.js [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+const util = require('util');
+
+const CTABUS_API_KEY = 'mY73pz65XVB4Yc7GYAgqFrHQY';
+const CTABUS_API_DOMAIN = 'http://ctabustracker.com';
+const CTABUS_API_PATH = '/bustime/api/v2/getstops';
+
+
+/*take latitude, long, return nearest stopID*/
+exports.getNearestBusStopId = (parameters, latitude, longitude, callback) => {
+    console.log("Entered GetNearestStopID Function:");
```
Comment (baumannalexj) [inline lambda/custom/repositories/transit/CtaBusRepository.js]: reminder

File: lambda/custom/repositories/transit/CtaBusRepository.js [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+        path: CTABUS_API_PATH,
+        queryParams: {
+            key: CTABUS_API_KEY,
+            rt: parameters.rt,
+            dir: parameters.dir,
+            format: "json"
+        }
+        
+    });
+    console.log(url);
```
Comment (baumannalexj) [inline lambda/custom/repositories/transit/CtaBusRepository.js]: reminder

File: lambda/custom/repositories/transit/CtaBusRepository.js [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+
+    request(url,  (error, response, body) => {      
+        console.log("CTABusRepository.getNearestStopId()");
```
Comment (baumannalexj) [inline lambda/custom/repositories/transit/CtaBusRepository.js]: reminder

File: lambda/custom/repositories/transit/CtaBusRepository.js [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+        console.log("API Return: \n" + body);
+        let stopId = closestStopId(latitude, longitude, JSON.parse(body));
+        callback(stopId);        
+    });
+};
+
+/* take JSON response from API, return nearest stopID*/
+closestStopId = (latitude, longitude, ctaJSONResponse) => {
+    console.log("Entered CtaBusRepository.closestStopID()");
```
Comment (baumannalexj) [inline lambda/custom/repositories/transit/CtaBusRepository.js]: reminder

File: lambda/custom/repositories/transit/CtaBusRepository.js [hunk 14]
```diff
@@ -0,0 +1,55 @@
...
+    var length = Object.keys(ctaJSONResponse["bustime-response"].stops).length;
+    console.log(length);
+    var closestStopID = ctaJSONResponse["bustime-response"].stops[0].stpid;
+    var closestDistance = distanceCalc.getDistance(latitude, longitude, ctaJSONResponse["bustime-response"].stops[0].lat, ctaJSONResponse["bustime-response"].stops[0].lon);
+    for (let i = 0; i < length; i++){
+        var lat = ctaJSONResponse["bustime-response"].stops[i].lat;
+        var long = ctaJSONResponse["bustime-response"].stops[i].lon;
+        var thisDistance = distanceCalc.getDistance(latitude,longitude,lat,long,'M');
+        var thisStopId = ctaJSONResponse["bustime-response"].stops[i].stpid
+        console.log("Distance to stop " + thisStopId + ": " + thisDistance);
```
Comment (baumannalexj) [inline lambda/custom/repositories/transit/CtaBusRepository.js]: reminder

File: lambda/custom/repositories/transit/CtaBusRepository.js [hunk 15]
```diff
@@ -0,0 +1,55 @@
...
+        if (thisDistance < closestDistance){
+            closestDistance = thisDistance;
+            closestStopId = thisStopId;
+        }
+    }
+    console.log("Closest StopID: " + closestStopId);
```
Comment (baumannalexj) [inline lambda/custom/repositories/transit/CtaBusRepository.js]: remidner

File: intent-service/skill.json [hunk 16]
```diff
@@ -0,0 +1,32 @@
+{
+  "manifest": {
+    "publishingInformation": {
+      "locales": {
+        "en-US": {
+          "name": "ChicagoNow-dan"
```
Comment (baumannalexj) [inline intent-service/skill.json]: I think this file is errant

**Review by baumannalexj — CHANGES_REQUESTED** [review]
we'll clean up the console outs and making constants

File: lambda/custom/index.js [hunk 17]
```diff
@@ -21,9 +22,17 @@ const handlers = {
...
     'CtaBusIntent': function () {
-        // TODO: Build proper parameters object
-
-        this.emit(':tell', "implement bus intent");
+        let transitSlot = this.event.request.intent.slots.transitMode.value;
+        if (transitSlot === "bus") {
+            let parameters = ParameterHelper.getLocationParameters(this.event.context.System);
+            parameters.rt = this.event.request.intent.slots.busStop.value;
+            parameters.dir = this.event.request.intent.slots.busDirection.value;
+            BusHandler.searchBusNearMe(parameters, (alexaResponse) => {
```
Comment (rknruben56) [inline lambda/custom/index.js:30]: shouldn't need BusHandler in index.js since TransitHandler has the method but we can always change it later if needed.

**Review by rknruben56 — APPROVED** [review]

**Pull request closed without merging**